httpcore/_synchronization.py

```diff
@@ -171,6 +171,58 @@ async def release(self) -> None:
             self._anyio_semaphore.release()
 
 
+class AsyncShieldCancellation:
+    # For certain portions of our codebase where we're dealing with
+    # closing connections during exception handling we want to shield
+    # the operation from being cancelled.
+    #
+    # with AsyncShieldCancellation():
+    #     ... # clean-up operations, shielded from cancellation.
+
+    def __init__(self) -> None:
+        """
+        Detect if we're running under 'asyncio' or 'trio' and create
+        a semaphore with the correct implementation.
+        """
+        self._backend = sniffio.current_async_library()
+
+        if self._backend == "trio":
+            if trio is None:  # pragma: nocover
+                raise RuntimeError(
+                    "Running under trio requires the 'trio' package to be installed."
+                )
+
+            self._trio_shield = trio.CancelScope(shield=True)
+        else:
+            if anyio is None:  # pragma: nocover
+                raise RuntimeError(
+                    "Running under asyncio requires the 'anyio' package to be installed."
+                )
+
+            self._anyio_shield = anyio.CancelScope(shield=True)
+
+    def __enter__(self) -> "AsyncShieldCancellation":
+        if not self._backend:
+            self.setup()
+
+        if self._backend == "trio":
+            self._trio_shield.__enter__()
+        else:
+            self._anyio_shield.__enter__()
```

Comment:
Why not just set `self._shield` in `__init__` since the interfaces are identical?

Reply:
1. Maybe?
2. Not sure that'd pass type-checking?
3. Consistency with other cases?

Reply:
I think it'd pass type checking especially if you annotated it with a `Union`. No strong preference here just an observation of multiple conditional checks on the backend when really the backend stops mattering once you are done with `__init__`.

I'd do whatever you've been doing elsewhere in httpcore!